### PR TITLE
More robust eta expansion check in old unification

### LINKED ISSUE #1614: Broken in coq CI
eg https://gitlab.com/coq/coq/-/jobs/3668196302
Seems to be due to https://github.com/UniMath/UniMath/pull/1610

### LINKED ISSUE #17252: Unification failure from unimath
cf https://github.com/UniMath/UniMath/issues/1614

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #17252
Fix https://github.com/UniMath/UniMath/issues/1614

The bug could not be minimized so no test, maybe autominification on a revert of this change will work better.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
HoTT case 2 further reduced
~~~coq


Module Export Datatypes.

Set Implicit Arguments.
Local Set Primitive Projections.

Record prod (A B : Type) := pair { fst : A ; snd : B }.

Notation "x * y" := (prod x y) : type_scope.
Notation "( x , y , .. , z )" := (pair .. (pair x y) .. z) : core_scope.

End Datatypes.


Delimit Scope morphism_scope with morphism.
Delimit Scope category_scope with category.
Delimit Scope object_scope with object.

Record PreCategory :=
  Build_PreCategory {
      object :> Type;
      morphism : object -> object -> Type;
      identity : forall x, morphism x x;
    }.

Bind Scope category_scope with PreCategory.
Arguments identity {!C%category} / x%object : rename.

Local Open Scope morphism_scope.

Section prod.
  Variables C D : PreCategory.

  Definition prod : PreCategory := (@Build_PreCategory
              (C * D)%type
              (fun s d => (morphism C (fst s) (fst d)
                           * morphism D (snd s) (snd d))%type)
              (fun x => (identity (fst x), identity (snd x)))).

End prod.

Local Infix "*" := prod : category_scope.
Delimit Scope functor_scope with functor.

Section Functor.
  Variables C D : PreCategory.

  Record Functor :=
    {
      object_of :> C -> D;
      morphism_of : forall s d, morphism C s d
                                -> morphism D (object_of s) (object_of d);
      identity_of : forall x, morphism_of _ _ (identity x)
                              = identity (object_of x)
    }.
End Functor.
Arguments morphism_of [C%category] [D%category] F%functor [s%object d%object] m%morphism : rename, simpl nomatch.

Variables C1 C2 D : PreCategory.

Variable F : Functor (C1 * C2) D.

Lemma foo (c1:C1) (x : object C2)
  : @morphism_of _ _ F
      (@pair (object C1) (object C2) c1 x)
      (@pair (object C1) (object C2) c1 x)
      (identity c1, identity x)
    = identity (F (c1, x)).
Proof.
  rewrite identity_of.
~~~

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coq/pretyper-maintainers should be ready

File: pretyping/unification.ml [hunk 1]
```diff
@@ -851,15 +851,8 @@ let rec unify_0_with_initial_metas (sigma,ms,es as subst : subst0) conv_at_top e
...
-             let opt' = {opt with at_top = true; with_cs = false} in
-             let substn =
-               let tM, tN =
-                 try get_type_of ~lax:true env sigma cM
-                   , get_type_of ~lax:true env sigma cN
-                 with Retyping.RetypeError _ -> error_cannot_unify env sigma (cM, cN)
-               in
-               unirec_rec curenvnb CONV opt' ~nargs:0 substn tM tN
-             in
+             let opt' = {opt with at_top = true; with_cs = true} in
```
Comment (maximedenes) [inline pretyping/unification.ml:854]: Shouldn't we keep `with_cs = false`?

**Comment by SkySkimmer** [comment]
@maximedenes merge?

**Review by maximedenes — APPROVED** [review]
LGTM

**Comment by maximedenes** [comment]
@coqbot merge now

**Pull request merged**